sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.students.controller.getInProgressApplicationDetails.e2e-spec.ts

```diff
@@ -0,0 +1,409 @@
+import { HttpStatus, INestApplication } from "@nestjs/common";
+import * as request from "supertest";
+import {
+  BEARER_AUTH_TYPE,
+  createTestingAppModule,
+  getStudentToken,
+  getStudentByFakeStudentUserType,
+  FakeStudentUsersTypes,
+  mockOutstandingAssessment,
+} from "../../../testHelpers";
+import {
+  createFakeCRAIncomeVerification,
+  createFakeSupportingUser,
+  createE2EDataSources,
+  E2EDataSources,
+  saveFakeApplication,
+  createFakeStudentAssessment,
+} from "@sims/test-utils";
+import {
+  Student,
+  StudentAssessmentStatus,
+  SupportingUserType,
+} from "@sims/sims-db";
+import { SuccessWaitingStatus } from "../models/application.dto";
+import { addDays, getISODateOnlyString } from "@sims/utilities";
+import { TestingModule } from "@nestjs/testing";
+
+describe("ApplicationStudentsController(e2e)-getApplicationProgressDetails", () => {
+  let app: INestApplication;
+  let db: E2EDataSources;
+  let appModule: TestingModule;
+  let student: Student;
+
+  beforeAll(async () => {
+    const { nestApplication, module, dataSource } =
+      await createTestingAppModule();
+    app = nestApplication;
+    db = createE2EDataSources(dataSource);
+    appModule = module;
+    student = await getStudentByFakeStudentUserType(
+      FakeStudentUsersTypes.FakeStudentUserType1,
+      dataSource,
+    );
+  });
+
+  it("Should throw not found error when application is not found.", async () => {
+    // Arrange
+
+    const endpoint = `/students/application/99999999/in-progress`;
+    const token = await getStudentToken(
+      FakeStudentUsersTypes.FakeStudentUserType1,
+    );
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(token, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.NOT_FOUND);
+  });
+
+  it("Should get application in-progress details when the student does not have a supporting user.", async () => {
+    // Arrange
+
+    const application = await saveFakeApplication(db.dataSource, { student });
+
+    // Create CRA income verifications for student.
+    const studentCRAIncomeVerification = createFakeCRAIncomeVerification(
+      {
+        application,
+      },
+      { initialValues: { dateReceived: new Date() } },
+    );
+    await db.craIncomeVerification.save([studentCRAIncomeVerification]);
+
+    // Mock assessment service to return the saved student assessment.
+    await mockOutstandingAssessment(appModule, application.currentAssessment);
```

Comment:
I did not follow why we have to do a mock here?

If there is no previous application, the current assessment of the application will be returned as first outstanding assessment.

Reply:
IMO, it is particularly important to use a mock method for the method `getOutstandingAssessmentsForStudentInSequence` in all the E2E test cases in this file. There are many student assessment records in the test database, and `getOutstandingAssessmentsForStudentInSequence` will return the earliest in-progress/submitted assessment for the program year with the earliest study start date for the offering. For that reason, each of these test cases here will fail if I didn't mock the result from the method `getOutstandingAssessmentsForStudentInSequence`; i.e., it will return a student assessment that isn't the current assessment created in the test case. This mock method ensures that the results from `getOutstandingAssessmentsForStudentInSequence` is well-controlled despite the database and the endpoint returns expected data. This is also in response to the comment below: https://github.com/bcgov/SIMS/pull/3208#discussion_r1591344481

Reply:
I see that there is some gap in understanding regarding the way we execute student E2E tests.

For student API tests  what we have been doing is

1. We can create a unique student for the test to avoid data conflicts.
2. Mock the user token value to return the unique student we created using `mockUserLoginInfo()`
3. Test for data related to the unique student.

Here is an example. 

![image](https://github.com/bcgov/SIMS/assets/54600590/2058e089-6b08-4e01-8d20-49212a7d65c1)